### DISCUSSION TITLE
DataGridCollectionView.SortDescriptions.CollectionChanged event stops firing after hosting TabControl changes tabs

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#14701] [Q&A]
I have a TabControl that is hosting a DataGrid.
I am subscribing to the SortDescriptions.CollectionChanged event of the DataGridCollectionView associated with the DataGrid.
The event fires if I change the sorting of the DataGrid by clicking on a column header.

However, after I switch to another tab (which fires the DataGrid.DataContextChanged event), the sorting on the first tab is lost, and moreover, the CollectionChanged event stops firing from then on, although the sorting does appear to occur.

How can I make the CollectionChanged continue to work?

**Comment by timunie** [comment]
I tend to wrap my ItemsSource in a DataGridCollectionView. However probably you need to unregister when the View gets detached from visual tree?

**Comment by wgraaf** [comment]
I can make the CollectionChanged  event work by wrapping the ItemsSource in a DataGridCollectionView in the DataGrid.Sorting event.

I would like to do that earlier, for instance in DataGrid.DataContextChanged, but this gets called BEFORE the ItemsSource gets set, 
so anything I do there get blown away.

Performing the ItemsSource wrap would also make it easier to retain the sorting on each tab when the user changes tabs.

What I am missing is an ItemsSource changed event.
Is there an event that I could use instead?

**Comment by wgraaf** [comment]
Is there an equivalent in Avalonia to DependencyPropertyDescriptors in WPF?

```
    DependencyPropertyDescriptor itemsSourcePropertyDescriptor = DependencyPropertyDescriptor.FromProperty(ItemsControl.ItemsSourceProperty, typeof(DataGrid));
    itemsSourcePropertyDescriptor.AddValueChanged(dataGrid1, OnDataGridItemsSourceChanged);

private void OnDataGridItemsSourceChanged(object sender, EventArgs e) { }
```
